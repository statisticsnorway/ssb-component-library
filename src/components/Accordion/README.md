Accordion
========

> Expandable accordion

### Usage

#### HTML

```html
<div class="ssb-accordion">
  <div class="accordion-header closed" onclick="{toggle classname to 'open'}">
    {plus icon}
    <div class="ssb-title header-text no-margin"><h5>Title</h5></div>
  </div>
  <div class="accordion-body closed">
    {insert content}
  </div>
</div>

<div class="ssb-nested-accordion mt-3">
  <div class="nested-accordion-header closed" onclick="{toggle classname to 'open'}">
    {plus icon}
    <span class="header-text">Title</span>
  </div>
  <div class="accordion-body closed">
    {insert content}
  </div>
</div>
```
Toggle the all instances of the `closed` className to `open` when clicked.

#### React

```jsx harmony
<Accordion primary header="This is a primary header">
  This is paragraph text which explains the accordion, the rest of the text is just to fill it out and show the space it takes.
</Accordion>

<Accordion primary header="This is a primary header">
  This is paragraph text which explains the accordion, the rest of the text is just to fill it out and show the space it takes.
  <NestedAccordion header="A nested accordion header">
    This is paragraph text which explains the accordion, the rest of the text is just to fill it out and show the space it takes.
  </NestedAccordion>
</Accordion>
```

Available props:

| Name       | Type           | Description  |
| ---------- | ------------- | ----- |
| children | node | Body content |
| header | string | Header text |
| openByDefault | bool | Open when rendered. Defaults to `false` |
| primary | bool | Changes design |
| subHeader | string | Adds a sub header |
