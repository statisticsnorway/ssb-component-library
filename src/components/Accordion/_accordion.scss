.ssb-accordion {
  display: inline-grid;
  width: 100%;

  .accordion-header {
    @include reset-button;

    .button-grid {
      align-items: center;
      cursor: pointer;
      display: grid;
      grid-column-gap: $global-padding-size;
      grid-template-columns: 20px auto 1fr;
    }

    .expand-icon {
      color: $ssb-white;
      fill: $ssb-green-3
    }

    .sub-header {
      color: $ssb-green-3;
    }

    &:hover {

      .header-text, .sub-header {
        color: $ssb-green-4;
      }
    }

    &.open {

      .header-text, .sub-header {
        color: $ssb-green-4;
      }

      .expand-icon {
        fill: $ssb-green-4;
      }
    }
  }

  &.with-sub-header {

    .accordion-header {
      grid-template-columns: 20px auto;

      .header-text {
        grid-column-start: 2;
        grid-row-start: 2;
      }
    }
  }

  .accordion-body {
    @include open-sans;
    font-size: 16px;
    padding: $global-padding-size $global-padding-size $global-padding-size 30px;

    &.closed {
      display: none;
      height: 0;
    }
  }
}
