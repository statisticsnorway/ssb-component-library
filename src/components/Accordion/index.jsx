import React, { useState } from 'react';
import PropTypes from 'prop-types';
import { PlusCircle, MinusCircle } from 'react-feather';
import TitleÂ from '../Title';

const Accordion = ({
	children, header, openByDefault, primary, subHeader,
}) => {
	const [isOpen, toggleOpen] = useState(openByDefault);
	return (
		<div className={`ssb-accordion ${subHeader && 'with-sub-header'}`}>
			<div
				className={`accordion-header ${isOpen ? 'open' : 'closed'}`}
				onClick={() => toggleOpen(!isOpen)}
			>
				{!isOpen && <PlusCircle className="expand-icon" size={primary ? 20 : 15} />}
				{isOpen && <MinusCircle className="expand-icon" size={primary ? 20 : 15} />}
				{subHeader && <Title size={5} className="sub-header no-margin">{subHeader}</Title>}
				<Title size={5} className="header-text no-margin">{header}</Title>
			</div>
			<div className={`accordion-body ${isOpen ? 'open' : 'closed'}`}>
				{isOpen && children}
			</div>
		</div>
	);
};

Accordion.defaultProps = {
	openByDefault: false,
	primary: false,
};

Accordion.propTypes = {
	children: PropTypes.node,
	header: PropTypes.string,
	openByDefault: PropTypes.bool,
	primary: PropTypes.bool,
	subHeader: PropTypes.string,
};

export default Accordion;
