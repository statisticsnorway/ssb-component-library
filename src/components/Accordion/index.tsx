import React, { useEffect, useRef, useState, type ReactNode } from 'react'
import { ChevronDown, ChevronUp } from 'react-feather'

interface AccordionProps {
  id?: string
  children?: ReactNode
  className?: string
  header?: string
  openByDefault?: boolean
  subHeader?: string
  tabIndex?: number
  withoutBorders?: boolean
  onToggle?: (isOpen: boolean) => void
}

const Accordion: React.FC<AccordionProps> = ({
  id,
  children,
  className,
  header,
  openByDefault = false,
  subHeader,
  tabIndex = 0,
  withoutBorders,
  onToggle = () => {},
}) => {
  const [isOpen, toggleOpen] = useState(openByDefault)
  const firstUpdate = useRef(true)
  useEffect(() => {
    if (firstUpdate.current) {
      firstUpdate.current = false
      return
    }
    onToggle(isOpen)
  }, [isOpen])
  return (
    <div
      id={id}
      className={`ssb-accordion${subHeader ? ' with-sub-header' : ''}${withoutBorders ? ' without-borders' : ''}${className ? ` ${className}` : ''}`}
    >
      <button
        className={`accordion-header ${isOpen ? 'open' : 'closed'}`}
        aria-expanded={isOpen ? 'true' : 'false'}
        tabIndex={tabIndex}
        onClick={() => toggleOpen(!isOpen)}
      >
        <span className='button-grid'>
          {subHeader && <span className='sub-header'>{subHeader}</span>}
          <span className='header-text'>{header}</span>
          {!isOpen && <ChevronDown className='expand-icon' size={20} />}
          {isOpen && <ChevronUp className='expand-icon' size={20} />}
        </span>
      </button>
      <div className={`accordion-body ${isOpen ? 'open' : 'closed'}`}>{children}</div>
    </div>
  )
}

export default Accordion
