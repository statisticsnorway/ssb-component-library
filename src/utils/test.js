import fs from 'fs';
import path from 'path';
import { render as rtlRender } from '@testing-library/react';

export * from '@testing-library/react';
export * from '@testing-library/jest-dom';

const styles = fs.readFileSync(path.relative(process.cwd(), './lib/bundle.css'), 'utf8');

export function render(ui, options) {
	const view = rtlRender(ui, options);

	const styleElement = document.createElement('style');
	styleElement.innerHTML = styles;
	document.body.appendChild(styleElement);

	return view;
}
