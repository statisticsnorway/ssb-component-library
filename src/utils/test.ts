import fs from 'fs'
import path from 'path'
import { RenderOptions, RenderResult, render as rtlRender } from '@testing-library/react'
import { ReactNode } from 'react'

export * from '@testing-library/react'

const styles = fs.readFileSync(path.relative(process.cwd(), './lib/bundle.css'), 'utf8')

export function render(ui: ReactNode, options?: RenderOptions): RenderResult {
  const view = rtlRender(ui, options)

  const styleElement = document.createElement('style')
  styleElement.innerHTML = styles
  document.body.appendChild(styleElement)

  return view
}
