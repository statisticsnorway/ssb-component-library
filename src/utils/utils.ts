// utils.ts
const classNames = (...classes: (string | boolean | undefined)[]): string => classes.filter(Boolean).join(' ')

export default classNames
