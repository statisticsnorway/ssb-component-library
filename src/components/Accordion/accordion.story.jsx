import React from 'react';
import { storiesOf } from '@storybook/react';
import centered from '@storybook/addon-centered/react';
import Accordion from './index';
import NestedAccordion from '../NestedAccordion';

const mockText = `This is paragraph text which explains the accordion,
 the rest of the text is just to fill it out and show the space it takes.`;

storiesOf('Accordion', module).addDecorator(centered)
	.add('Default', () => (
		<div style={{ display: 'flex', width: '500px' }}>
			<Accordion header="This is a accordion title">
				This is paragraph text which explains the accordion, the rest of the text is just to fill it out and show the space it takes.
			</Accordion>
		</div>
	))
	.add('With sub-header', () => (
		<div style={{ display: 'flex', width: '500px' }}>
			<Accordion subHeader="Tabell 1" header="This is a accordion title with a sub-header">
				This is paragraph text which explains the accordion, the rest of the text is just to fill it out and show the space it takes.
			</Accordion>
		</div>
	))
	.add('With nested accordions', () => (
		<div style={{ display: 'flex', width: '500px' }}>
			<Accordion header="This is a accordion with nested accordions">
				<div style={{ marginBottom: '10px' }}>{mockText}</div>
				<NestedAccordion header="A nested accordion header">
					This is paragraph text which explains the accordion, the rest of the text is just to fill it out and show the space it takes.
				</NestedAccordion>
			</Accordion>

		</div>
	))
	.add('Without borders', () => (
		<div style={{ display: 'flex', width: '500px' }}>
			<Accordion withoutBorders header="This is a accordion title without borders">
				This is paragraph text which explains the accordion, the rest of the text is just to fill it out and show the space it takes.
			</Accordion>
		</div>
	))
	.add('Multiple w/ sub-header', () => (
		<div style={{ display: 'flex', flexDirection: 'column', width: '500px' }}>
			<Accordion subHeader="Tabell 1" header="This is a accordion title with a sub-header, her you can see the text in an extra row width a 6 column max lenght">
				This is paragraph text which explains the accordion, the rest of the text is just to fill it out and show the space it takes.
			</Accordion>
			<Accordion subHeader="Tabell 2" header="This is a accordion title with a sub-header">
				This is paragraph text which explains the accordion, the rest of the text is just to fill it out and show the space it takes.
			</Accordion>
			<Accordion subHeader="Tabell 3" header="This is a accordion with nested accordions">
				<NestedAccordion header="A nested accordion header">
					This is paragraph text which explains the accordion, the rest of the text is just to fill it out and show the space it takes.
				</NestedAccordion>
			</Accordion>
		</div>
	));
