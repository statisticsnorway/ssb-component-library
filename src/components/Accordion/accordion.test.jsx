import React from 'react'
import userEvent from '@testing-library/user-event'

import { screen, render } from '../../utils/test'
import Accordion from './index'

const headerText = 'This is a accordion title'
const bodyText = 'This is paragraph text'

describe('Accordion component', () => {
  test('Matches the snapshot', () => {
    const { asFragment } = render(<Accordion header={headerText}>Accordion</Accordion>)
    expect(asFragment()).toMatchSnapshot()
  })
  test('Starts closed', async () => {
    render(<Accordion header={headerText}>{bodyText}</Accordion>)
    expect(await screen.findByText(bodyText)).not.toBeVisible()
  })
  test('Open on click when starting closed', async () => {
    const user = userEvent.setup()
    render(<Accordion header={headerText}>{bodyText}</Accordion>)

    const button = await screen.findByText(headerText)
    await user.click(button.parentElement)

    expect(await screen.findByText(bodyText)).toBeVisible()
  })
  test('Accordion with sub-header', async () => {
    const subheader = 'Subheader'
    render(
      <Accordion header={headerText} subHeader={subheader}>
        {bodyText}
      </Accordion>
    )
    expect(await screen.findByText(subheader)).toBeVisible()
  })
})
