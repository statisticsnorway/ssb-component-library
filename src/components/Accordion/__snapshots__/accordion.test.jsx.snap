// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Accordion component Matches the snapshot 1`] = `
<DocumentFragment>
  <div
    class="ssb-accordion"
  >
    <button
      aria-expanded="false"
      class="accordion-header closed"
      tabindex="0"
    >
      <span
        class="button-grid"
      >
        <span
          class="header-text"
        >
          This is a accordion title
        </span>
        <svg
          class="expand-icon"
          fill="none"
          height="20"
          stroke="currentColor"
          stroke-linecap="round"
          stroke-linejoin="round"
          stroke-width="2"
          viewBox="0 0 24 24"
          width="20"
          xmlns="http://www.w3.org/2000/svg"
        >
          <polyline
            points="6 9 12 15 18 9"
          />
        </svg>
      </span>
    </button>
    <div
      class="accordion-body closed"
    >
      Accordion
    </div>
  </div>
</DocumentFragment>
`;
